[MM-47326]: Contribute - WYSIWYG editor

michelengelen opened the pull request:
#### Summary
This will add the WYSIWYG editor to the webapp.

For now this is behind a feature flag and a setting in the advanced section since the editor as it is now is pretty much in an alpha state. In an effort to keep working on this in smaller iterations to be more effective the team decided we can release this as is for now and deal with any bugs as we go on.

to enable it you have to set the feature flag: `MM_FEATUREFLAGS_WYSIWYGEDITOR=true`, then the setting in the advanced settings will appear and you can use it in the create post section

#### Ticket Link
EPIC: [MM-47326](https://mattermost.atlassian.net/browse/MM-47326)

#### Related Pull Requests
- Has server changes: https://github.com/mattermost/mattermost-server/pull/22020

This is a re-opened PR from #11186 

#### Release Note
```release-note
Adds the alpha state of the WYSIWYG editor to the create post section behind feature flag and advanced setting
```


[MM-47326]: https://mattermost.atlassian.net/browse/MM-47326?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jfrerich committed:
initial add of advanced user setting

michelengelen committed:
WIP

michelengelen committed:
WIP

jfrerich committed:
[contributor / wysiwyg] Add Feature Flag (#11185)

michelengelen committed:
WIP

michelengelen committed:
Merge remote-tracking branch 'origin/contribute/wysiwyg' into contribute/wysiwyg

michelengelen committed:
WIP

michelengelen committed:
WIP

michelengelen committed:
added basic select to the toolbar with header options and tidied up a bit

michelengelen committed:
WIP

hmhealey committed:
Add draft draft support

michelengelen committed:
added basic send post functionality

michelengelen committed:
added heading formatting

michelengelen committed:
added code language select

michelengelen committed:
- enabled settings guarding
- set new autofocus property
- fixed storing the draft

michelengelen committed:
WIP

michelengelen committed:
fixed user-setting

michelengelen committed:
added basic table support

michelengelen committed:
added more sophisticated table support

michelengelen committed:
finished cleanup, restructuring and i18n support

michelengelen committed:
added new table control styling

michelengelen committed:
changed some naming and added correct translations

michelengelen committed:
some changes around tables

michelengelen committed:
added user and channel suggestions

michelengelen committed:
added emoji suggestion

michelengelen committed:
added basic overlay support for links

michelengelen committed:
restructured files a bit

michelengelen committed:
added hashtag basic support

michelengelen committed:
added link support

michelengelen committed:
added image upload files to use it for generic file-upload

michelengelen committed:
adde delete icon and correct helper text to link overlay

michelengelen committed:
custom react renderer for the atmentions

michelengelen committed:
restructured atmention-suggestion as an example for the others

michelengelen committed:
added missing items to the atmention suggestions

michelengelen committed:
fixed rendering of actual mention item in editor

michelengelen committed:
at mentions now work as expected

michelengelen committed:
added channel suggestion to the WYSIWYG editor

michelengelen committed:
removed turndown since we dont use that anymore

michelengelen committed:
got emoji suggestion working

michelengelen committed:
adjusted emoji name display

michelengelen committed:
adjusted emoji name display in rendered HTML

michelengelen committed:
added license guard to group mentions

michelengelen committed:
added correct enter behavior and connected to redux from index file

michelengelen committed:
added correct enter behavior and connected to redux from index file

michelengelen committed:
fixed some review findings from UX

michelengelen committed:
slightly changed link behavior

michelengelen committed:
- added support for slash commands
- adjusted behavior of pressing '(mod/shift)+Enter' when suggestions are active
- added some comments

michelengelen committed:
adjusted styling of command suggestion items

michelengelen committed:
added working emojipicker (positioning tbd)

michelengelen committed:
fixed emojiPicker positioning
- gifPicker is currently not inserting the correct value

michelengelen committed:
- fixed unresponsive app when emojipicker is open
- added image extension and image rendering in editor
- troubleshooting markdown parsing

michelengelen committed:
things

michelengelen committed:
custom extensions

michelengelen committed:
- removed unfinished refactoring
- added missing shortcut overlays to toolbar controls

michelengelen committed:
added a few things from PR review

michelengelen committed:
restructure folders and extensions

michelengelen committed:
more restructuring and externalisation

michelengelen committed:
added additional controls support

michelengelen committed:
Merge branch 'master' into contribute/wysiwyg

# Conflicts:
#	components/advanced_create_comment/advanced_create_comment.tsx
#	components/advanced_create_post/advanced_create_post.tsx
#	components/advanced_create_post/index.ts
#	components/user_settings/advanced/index.ts
#	components/user_settings/advanced/user_settings_advanced.tsx
#	i18n/en.json
#	package-lock.json
#	package.json
#	packages/mattermost-redux/src/selectors/entities/posts.ts
#	packages/mattermost-redux/src/selectors/entities/preferences.ts
#	packages/types/src/utilities.ts

michelengelen committed:
- fixed issues wth npm packages after merging master
- added additionalControls

michelengelen committed:
added edit last post functionalty

michelengelen committed:
added config option to allow fine-tuning of formatting options

michelengelen committed:
improved configuration options for WYSIWYG editor

michelengelen committed:
fixed table selection UI

michelengelen committed:
adressed PR review comments

michelengelen committed:
Merge branch 'master' into contribute/wysiwyg

# Conflicts:
#	utils/emoji_map.ts

michelengelen committed:
fixed merge conflict

michelengelen committed:
added proprietary fileuploads to the WYSIWYG editor

phoinixgrr committed:
Empty-Commit

michelengelen committed:
fixed build
- added file- and uploadinfos to the drafting

michelengelen committed:
Merge remote-tracking branch 'origin/contribute/wysiwyg' into contribute/wysiwyg

michelengelen committed:
added some inline documentation

michelengelen committed:
Empty-Commit-test

michelengelen committed:
fix i18n check

michelengelen committed:
make eslint happy :D

michelengelen committed:
Merge branch 'master' into contribute/wysiwyg

michelengelen committed:
fixing linters and i18n

michelengelen committed:
fixed type linting (for now)

michelengelen committed:
added kaomoji picker as a fun project! :D

michelengelen committed:
added tracking to the advanced setting option

michelengelen committed:
Merge branch 'master' into contribute/wysiwyg

# Conflicts:
#	components/advanced_create_post/advanced_create_post.tsx
#	i18n/en.json

michelengelen committed:
removed unused string from translation file

matthewbirtch reviewed (changes requested):
@michelengelen this is awesome. A few more issues I found while testing today:

1. The input field's non-focused state doesn't seem to be right. It looks like we lost this state from the input. When not focused, the border should 16% opacity, and only when focused should it be 32%.
2. It doesn't seem to be honoring the setting for command-enter in `Settings > Advanced`. I have this set to 'On for all messages', but when I press enter it sends instead of adding a new line.

<img width="584" alt="image" src="https://user-images.githubusercontent.com/2040554/211652921-1b93003b-5458-472e-bc46-6078282a6193.png">

3. As mentioned previously, the Command-option-K shortcut for inserting a link is not working
4. Still need to implement the upload shortcut (Command-U)
5. The strikethrough formatting option should show next to the italic button
6. The shortcut for strikethrough (Shift-Option-X) doesn't seem to be working
7. The shortcut for code snippet (Command-Option-C) doesn't seem to be working
8. The shortcut for code block is showing in the tooltip as Command-Option-C which appears to be the same as code snippet. I guess we need to choose which one gets this shortcut and provide an alternate for one of them? In the product currently, Command-Option-C is used for the code snippet, so we should find an alternate for code blocks (or remove it for now until we can come up with another option)
9. In the link popover, I'm now seeing two icons that represent ENTER.
    <img width="434" alt="image" src="https://user-images.githubusercontent.com/2040554/211659328-a392c521-7ebe-4812-8f9f-974188965b4f.png">

10. While inside a code block, pressing down on the last line of code is no longer adding a new line outside the code block. Similarly, pressing up on the first line of code is no longer moving the cursor above the code block.
11. When selecting text on the right side and clicking the link button, the link popover opens too far on the right side and gets cut off
    <img width="1168" alt="image" src="https://user-images.githubusercontent.com/2040554/211660522-38c3d90b-95be-4591-b4c4-83ca976e3cfa.png">
12. In the emoji autocomplete, I'm seeing `SUGGESTION.MENTION.EMOJI` as the title which I assume is a translation ID.
    <img width="335" alt="image" src="https://user-images.githubusercontent.com/2040554/211661049-8bda70a6-3296-4b6e-b507-c28392a8aa7e.png">
13. When I open the emoji picker (by clicking the emoji icon on the right side), the `ESC` key does not close the picker.
14. Adding attachments to the input field works, but when I post the message, the attachment doesn't end up posting with the message.
15. Similarly, I am not able to post a message that only has an attachment and not text - the send button is disabled

### Non-blocking comments
- Should we consider linking `@mentions` and formatting them as links to be true WYSIWYG
- The send button needs a hover state. It seems like it lost its hover state (or it never had one?)
- In code blocks, the language label isn't quite aligned properly with the chevron
    <img width="89" alt="image" src="https://user-images.githubusercontent.com/2040554/211659829-650ae24e-140d-439a-88f0-c8c37f21d514.png">
- The size of emojis looks to be slightly different with the WYSIWYG editor vs. what gets rendered in the post
    <img width="176" alt="image" src="https://user-images.githubusercontent.com/2040554/211661750-cde1284a-bb24-4a4d-bd97-4a0de771dcdb.png">


### To be addressed separately
- add WYSIWYG editor to edit post
- add WYSIWYG editor the RHS and threads view reply input
- other places WYSIWYG could be added
- markdown parsing on paste

michelengelen committed:
fixes focussed border color of input

michelengelen committed:
fixed honoring the sending with mod-enter user-setting

michelengelen committed:
implemented upload hotkey

michelengelen committed:
a few smaller remarks

michelengelen committed:
fixed upload behavior

michelengelen committed:
fixed linter and i18n CI pipelines

michelengelen committed:
fixed linter and i18n checks

matthewbirtch commented:
Looks like you fixed most of the issues @michelengelen. 

A couple more things:
1. For the enter icon, can we adjust the alignment a bit and add a few px of space between the icon and the word ENTER?
<img width="111" alt="image" src="https://user-images.githubusercontent.com/2040554/211956024-2e450b10-d2ed-458c-ae60-fcadd13e3d7f.png">

2. Also on smaller screens, we're not collapsing any formatting buttons by the looks of it like we do with the current advanced text editor. Things seem to be bleeding out of the container. Is this something you'll work on in a separate PR?
<img width="1022" alt="image" src="https://user-images.githubusercontent.com/2040554/211956274-70b5a484-b313-43f3-8b41-1eff0432872b.png">

michelengelen committed:
- fixed codeBlock exit behavior
- fixed misalignment in link overlay helper text
- should fix i18n pipeline
- fixed emoji size in editor

michelengelen committed:
finally fixes i18n (hopefully)

michelengelen commented:
@matthewbirtch same procedure as last time:
 
>     1. For the enter icon, can we adjust the alignment a bit and add a few px of space between the icon and the word ENTER?

Fixed!

>     2. Also on smaller screens, we're not collapsing any formatting buttons by the looks of it like we do with the current advanced text editor. Things seem to be bleeding out of the container. Is this something you'll work on in a separate PR?

This is a bit hard to do without excessive calculations. I would like to try and find the best solution possible. I will take this on in the second iteration (same as the hidden formatting bar feature)

On a side note: 
1. I fixed the `ArrowDown` behavior in code blocks. Still missing the `ArrowUp` behavior though, which is a lot harder to do
2. I fixed the emoji sizing in the editor ... should now be the same as in posts (except for emoji posts)

michelengelen committed:
- fixed unsetting link via button
- added wide and normal mode for the toolbar

michelengelen committed:
extracted submitHandler in linkoverlay

michelengelen committed:
added WYSIWYG editor to thread view

michelengelen committed:
fixed file uploads in comment wysiwyg

michelengelen committed:
tried adding to the edit post view

esethna commented:
@michelengelen really great work, inot sure how many of these have already been found but here's a few things I noticed:

1. When selecting a code block, it defaults to CSS (can we default to no specific language selected but still have the dropdown available for a language selection (may need a placeholder text?)?

2. The language selector might be broken:
![image](https://user-images.githubusercontent.com/14333569/212195885-65eecbf3-f45a-489f-b6f2-a238775d2bae.png)

3. Message priority button doesn't seem to be working

4. When tables are used, can we allow rows & columns in the table, followed by the delete key to delete the selected row/column? I had a hard time figuring out how to remove a column.

5. With tables, can we allow ctrl + A or a dragging across the table to select it for deletion. I found it a little finicky to figure out to properly select the table so it could be deleted

6. I think this was discussed already, but it would be really excellent if the table pop-over could be shown in context with the table when a selection is made in the table. 

7. After using the link button and entering a link, can we move the cursor to after the word (see google docs)

8. Any change we can implement this improvement along with the new editor? ;) https://mattermost.atlassian.net/browse/MM-48061

michelengelen committed:
adjusted styling for compact heading style mode

michelengelen committed:
fixed language autodetection in code block

michelengelen committed:
added auto-detection of pasted language to codeblock extension

michelengelen committed:
decoupled setting message in state on every change to prevent rerenders

michelengelen committed:
changed Link behavior

michelengelen commented:
@esethna Comments between the lines.

>     1. When selecting a code block, it defaults to CSS (can we default to no specific language selected but still have the dropdown available for a language selection (may need a placeholder text?)?

FIXED
 
>     2. Language selector should validate languages: https://gyazo.com/c04d68ef2a983a44e39ae1a3c92e9796

FIXED

>     3. The language selector needs to be allowed to span outside the input area:

FIXED

>     4. Message priority button doesn't seem to be working

FIXED

>     5. When tables are used, can we allow rows & columns in the table, followed by the delete key to delete the selected row/column? I had a hard time figuring out how to remove a column.

The whole table feature is still just very basic and the proposed changes are particularly hard to implement. For this we would need to render the whole structure differently. We want to provide a similar experience as with the editor in JIRA/Confluence, but as this is such a big task I would like to push this to a later iteration.

>     6. With tables, can we allow ctrl + A or a dragging across the table to select it for deletion. I found it a little finicky to figure out to properly select the table so it could be deleted

see comment on point 5.

>     7. I think this was discussed already, but it would be really excellent if the table pop-over could be shown in context with the table when a selection is made in the table.

see comment on point 5.

>     8. After using the link button and entering a link, can we move the cursor to after the word (see google docs)

FIXED

>     9. Any chance we can implement this improvement along with the new editor? ;) https://mattermost.atlassian.net/browse/MM-48061

This is also very hard to do since the trigger functionality has no way of knowing the current state of the editor(s) which would be needed for it. There are a few ways of doing this, but it is definitely out of scope for this PR

Let me know what you think

matthewbirtch commented:
Seems like there might be something wrong with the menu for code block language selection. The height is very short and can't see any items within it:
<img width="309" alt="image" src="https://user-images.githubusercontent.com/2040554/212718036-c3e2cd2c-5111-4584-aef8-18740c28f371.png">

michelengelen committed:
fixed typing issue

michelengelen commented:
@matthewbirtch this should be fixed by now ... the test server did not update and that's why it is currently broken

michelengelen commented:
@matthewbirtch @esethna the test server is updated with the newest commit

jgilliam17 commented:
Thanks @michelengelen 
Seeing few issues 
1. Attachments remain in textbox after posting
<img width="1212" alt="Screen Shot 2023-01-17 at 12 36 24 PM" src="https://user-images.githubusercontent.com/52937121/212999554-52263a8e-b79f-4348-bbbd-668ac2e898ed.png">

2. Emoji picker is displaying off screen in Safari. Hidden when RHS is open
<img width="1204" alt="Screen Shot 2023-01-17 at 2 54 04 PM" src="https://user-images.githubusercontent.com/52937121/212999740-9a95a093-12f8-48f0-81d0-d7ecba9d8d5b.png">

3. Safari: Cursor missing when options other than Normal Text are toggled 
4. Existing posts appear to shift down when typing or deleting text


5. Code language selector is too tall, top items are out of view
<img width="890" alt="Screen Shot 2023-01-17 at 3 15 19 PM" src="https://user-images.githubusercontent.com/52937121/213002409-0e8b86d2-31bb-4fdf-ab18-e691dd724260.png">

6. When pasting code block, it is not recognized and formatted ( this may be done later). I'm unable to get out of code formatting and add normal text after pasting something similar to screenshot, can't start a new line
<img width="302" alt="Screen Shot 2023-01-17 at 3 25 38 PM" src="https://user-images.githubusercontent.com/52937121/213005357-4d6cb0ee-a843-4e0e-9fc4-aff3784ceb83.png">

7. Horizontal cursor when table is present

<img width="210" alt="Screen Shot 2023-01-17 at 3 33 39 PM" src="https://user-images.githubusercontent.com/52937121/213006696-7dbc4961-d5c4-4f58-83ee-cf23c458cb31.png">

8. Table input cursor jumps from top cells to bottom right when user stops typing

https://user-images.githubusercontent.com/52937121/213007669-079157b5-c833-4469-a93f-4e9978f3d700.mov

9. Unable to turn linking off e.g. add link and hit enter, add space and try to toggle link off to add more text
10.  Safari: Unable to add link, hitting Enter while on the link popover doesn't do anything 
11. There is slight downward shift to main channel posts (existing) when opening post priority menu. I have seen something similar with some other actions/posts. Maybe the same as issue 4.

https://user-images.githubusercontent.com/52937121/213014071-91864432-7c19-49d3-b404-0a858e08e950.mov

michelengelen committed:
fixes misaligned emoji picker

michelengelen committed:
moved post priority into the wysiwyg to prevent rerendering issues

michelengelen committed:
fixed max height on code selection menu

michelengelen committed:
fixed not working enter key on link popover

michelengelen committed:
added extend call to table to add markdown__table class to the table extensions rendering

michelengelen commented:
Hi @jgilliam17 ... thanks for going over this. Comments between the lines as usual:

>     1. Attachments remain in textbox after posting

FIXED

>     2. Emoji picker is displaying off screen in Safari. Hidden when RHS is open

FIXED

>     3. Safari: Cursor missing when options other than Normal Text are toggled

not sure what is causing this ... did not manage to fix for this iteration. @jgilliam17 could you open a ticket in the epic and assign this to me so I can work on this in the second iteration? Thanks!

>     4. Existing posts appear to shift down when typing or deleting text

FIXED

>     5. Code language selector is too tall, top items are out of view

FIXED (applied the Suggestion lists max-height value here - @matthewbirtch is that OK for you?)

>     6. When pasting code block, it is not recognized and formatted ( this may be done later). I'm unable to get out of code formatting and add normal text after pasting something similar to screenshot, can't start a new line

formatting pasted text (at least the more complex stuff) will be handled in one of the next iterations.

>     7. Horizontal cursor when table is present

This is the gapcursor and allows for easy deletion of the table when hitting Backspace

>     8. Table input cursor jumps from top cells to bottom right when user stops typing

I could not reproduce this (tested in FF, FF Dev Edition, Chrome and Safari)

>     9. Unable to turn linking off e.g. add link and hit enter, add space and try to toggle link off to add more text

You can exit the link with the Right-Arrow. This is the default behavior and since chars like dots and commas are valid chars for a link it is hard to determine when we want to exit the link state. For spaces: This is valid for text that has a link applied, so when you have a pure URL like `http://www.mattermost.com` it can be exited with a space.

>     10. Safari: Unable to add link, hitting Enter while on the link popover doesn't do anything

FIXED

>     11. There is slight downward shift to main channel posts (existing) when opening post priority menu. I have seen something similar with some other actions/posts. Maybe the same as issue 4.

FIXED (same issue as no.4)

> Edited: 12. Console error and banner. I am not 100% sure of the steps, but I think I had some text highlighted or pasted, follow by selecting to add table (happened on Desktop daily build)

Could not reproduce ... If you can reproduce and know the steps please open up a ticket.

Thanks for the review... Ready for another round! 🚀

matthewbirtch commented:
> > ```
> > 5. Code language selector is too tall, top items are out of view
> > ```
> 
> 
> FIXED (applied the Suggestion lists max-height value here - @matthewbirtch is that OK for you?)

Yep, this is good now

matthewbirtch commented:
> > ```
> > 7. Horizontal cursor when table is present
> > ```
> 
> 
> This is the gapcursor and allows for easy deletion of the table when hitting Backspace

@michelengelen I don't know if this is a common behavior - it was unexpected for me as well. I would have expected it just to be a normal text cursor on a new line (like we do for code blocks)

michelengelen commented:
> > > ```
> 
> > > 7. Horizontal cursor when table is present
> 
> > > ```
> 
> > 
> 
> > 
> 
> > This is the gapcursor and allows for easy deletion of the table when hitting Backspace
> 
> 
> 
> @michelengelen I don't know if this is a common behavior - it was unexpected for me as well. I would have expected it just to be a normal text cursor on a new line (like we do for code blocks)

I have seen this in a few other wysiwyg editors. I will Check if a normal caret is possible as well. This is the standard behavior for the package we are using.

jgilliam17 commented:
Thanks @michelengelen for the quick fix
Looks good, for alpha release to community
Noticed one small issue with Upload limit error, missing red font color on main channel, not reproducing on RHS.
<img width="896" alt="Screen Shot 2023-01-19 at 1 23 18 PM" src="https://user-images.githubusercontent.com/52937121/213533649-3fa8a4b5-a843-4614-9d23-b10ddb07c508.png">

michelengelen committed:
added missing 'MsgTyping' component and fixed error displays

michelengelen commented:
@jgilliam17 I fixed the error text color and added the `[user] is typing...` element as well.

@matthewbirtch the gapcursor is currently needed since when removing it from the extensions we cannot exit a table anymore when there is no text around it. I will leave it in for now, so we can see if this is acceptable in daily usage, or find another solution.

Thanks for your reviews!

Now its time for the devs to rip apart my code! 🤔

@mickmister @koox00 @AshishDhama and @Willyfrog: There are a few functions inside the `advanced_create_post.tsx` file that I have recreated to fit with the new data-flow. some just extract the message value from the event, where others are getting the string from the wysiwyg and need to loop that through (similar to the "old" implementations).

HAVE FUN!!!

matthewbirtch commented:
> added the `[user] is typing...` element as well.

The user is typing element is too small, but I believe that was going to be fixed in another PR since we identified that elsewhere, right?

matthewbirtch reviewed (approved):
Approving this for alpha release (on community) knowing there are polish and follow-up tickets/PRs to come. Awesome work @michelengelen

A few items to add to your list of follow-ups:
- I noticed that in the editor I can add a code block within a table cell. We should disable code blocks within tables since they won't properly render within the cell when posted anyway
- Same goes for bulleted lists, numbered lists, and quotes. They don't seem to render properly when added to table cells

jgilliam17 commented:
@michelengelen  additional note on `[user] is typing...` element is not visible for me on main channel. It is visible on RHS

AshishDhama commented:
@michelengele multiple checks failing, could please update the pr

michelengelen committed:
should fix i18n check

michelengelen committed:
should fix type-check pipeline

michelengelen committed:
fixed editor import

michelengelen committed:
fixed editor import

michelengelen committed:
removed index file

michelengelen committed:
fix linter

michelengelen commented:
@jgilliam17 strange ... its the same component ... could you try retesting it?

hmhealey commented on package.json:
```diff
@@ -88,6 +136,10 @@
     "redux-persist": "6.0.0",
     "redux-thunk": "2.4.2",
     "regenerator-runtime": "0.13.10",
+    "rehype-stringify": "9.0.3",
```
This is an awful lot of extra dependencies for just generating Markdown. I'm going to need to look through those in some more detail, but are you sure they're all needed?

michelengelen replied:
the problem is that we need to go through a few steps to convert the data in the editor to MD.

HTML => HAST => MDAST => MD 

and the other way around. Although I am currently not using that functionality

michelengelen replied:
fixed

hmhealey commented on package.json:
```diff
@@ -20,6 +20,39 @@
     "@stripe/react-stripe-js": "1.13.0",
     "@stripe/stripe-js": "1.41.0",
     "@tippyjs/react": "4.2.6",
+    "@tiptap/core": "2.0.0-beta.209",
```
Could we perhaps use Tiptap's StarterKit package to reduce the number of extensions here? https://tiptap.dev/api/extensions/starter-kit

michelengelen replied:
we could, but there is a ton of stuff in there we don't really need ... I kind of copied their approach and added everything they did add, but wanted to try and remove the stuff we don't need later on

hmhealey commented on packages/mattermost-redux/src/selectors/entities/preferences.ts:
```diff
@@ -268,6 +268,25 @@ export function getHasDismissedSystemConsoleLimitReached(state: GlobalState): bo
 }
 
+export function getWysiwygPreference(state: GlobalState): 'true' | 'false' {
+    return get(
```
This looks like it could use `getBool` instead of returning a string

hmhealey commented on packages/types/tsconfig.json:
```diff
@@ -14,5 +14,6 @@
     },
-    "exclude": ["**/node_modules", "**/lib", "**/*.test.js", "**/*.test.ts"]
+    "exclude": ["**/node_modules", "**/lib", "**/*.test.js", "**/*.test.ts"],
+    "typeRoots": ["node_modules/@types"],
```
What is this needed for?

michelengelen replied:
good question ... thats none of my changes ... might be a broken merge

hmhealey commented on reducers/drafts.ts:
```diff
@@ -0,0 +1,78 @@
+}
+
+export default combineReducers({
+    byChannel: persistReducer({key: 'drafts.byChannel', storage: localForage}, byChannel),
+    byThread: persistReducer({key: 'drafts.byThread', storage: localForage}, byThread),
```
Awesome, I've wanted to do this for ages. Having drafts stored in `state.storage` has actually caused quite a few performance issues in the past, so getting them stored in their own section of the state is awesome.

Does this work with the global drafts view and server-side syncing of drafts? Does it also support migrating old drafts to the new location/format on upgrade?

michelengelen replied:
thanks for the compliment, but this as well is not my change ... it has to be a broken merge here

hmhealey commented on sass/components/_channel-switcher.scss:
```diff
@@ -50,7 +50,9 @@
     .channel-switcher__suggestion-box {
         position: relative;
-        height: 362px;
+        height: 100%;
```
Is this change supposed to be part of this PR?

michelengelen replied:
same here

hmhealey commented on tsconfig.json:
```diff
@@ -23,6 +23,7 @@
     "baseUrl": ".",
     "useUnknownInCatchVariables": false,
     "paths": {
+        "@tiptap/core": ["node_modules/@tiptap/core/src/index"],
```
Is this needed? I wouldn't expect this sort of import rewriting to be needed for this library. The fact that we do it for the other packages is something specific to the monorepo

michelengelen replied:
it was needed, but apparently they fixed it with the new version

michelengelen replied:
Nope, they did not ... the types get not resolved correctly and therefor the `check-types` pipeline is failing

hmhealey commented on utils/constants.tsx:
```diff
@@ -2209,6 +2214,13 @@ export const InsightsCardTitles = {
     },
 };
 
+export enum WysiwygPluginNames {
```
Do these need to live here or is there somewhere they could closer to the WYSIWYG editor itself? It looks like they're only used within that folder, so I'd like to see them moved there

hmhealey commented on webpack.config.js:
```diff
@@ -139,6 +139,12 @@ var config = {
                 ],
             },
+            {
+                test: /\.m?js/,
```
What is this change for? There's a few things I notice that are potentially incorrect here:
1. This can capture plain old `.js` files because the `m` is optional which seems like we wouldn't want to have happen
2. Similarly, it can also capture `.jsx` files because it doesn't check for the end of the string with `$`

michelengelen replied:
fixed

hmhealey commented on components/emoji_picker/components/emoji_picker_kaomoji.tsx:
```diff
@@ -0,0 +1,205 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {ChangeEvent, KeyboardEvent, useEffect, useRef, useState} from 'react';
```
Could you pull this out and submit it as a separate PR? I'm 0/5 on if we add it, but this PR's already massive, so we should try to pare it down wherever possible

hmhealey reviewed (changes requested):
I haven't had a chance to look through the components themselves yet with this, but here's a review of the more infrastructure parts of it at least

jgilliam17 commented:
> @jgilliam17 strange ... its the same component ... could you try retesting it?

I checked today again and still miss the `user typing` on main channel

michelengelen commented on package.json:
```diff
@@ -51,16 +86,29 @@
     "popper.js": "1.16.1",
     "process": "0.11.10",
     "prop-types": "15.8.1",
+    "prosemirror-dropcursor": "1.6.1",
```
For some reason the peerDependencies do not work/are not installed when not adding them manually

michelengelen committed:
adds wysiwyg test file

michelengelen committed:
changed imports

michelengelen committed:
trying to fix tests

michelengelen committed:
made the tests work :D

michelengelen committed:
added snapshot for WYSIWYG

michelengelen committed:
added contentParser and respective tests

michelengelen committed:
Merge branch 'master' into contribute/wysiwyg

# Conflicts:
#	i18n/en.json
#	package-lock.json

michelengelen committed:
review remarks

michelengelen committed:
removed KaomojiPicker

michelengelen committed:
removed Kaomoji definitions

michelengelen committed:
fix type linting

michelengelen committed:
updated failing snapshot

michelengelen commented:
@hmhealey I have fixed the tests, removed the packages causing the problems, added a custom parser logic and included some tests for it as well. I did remove the Kaomoji-Stuff from this PR as well and will add a new PR in the following days!

@AshishDhama @mickmister @Willyfrog and @koox00 all tests are green now, so feel free to start reviewing

@jgilliam17 I will fix that `[user] is typing ...` issue next week.

Thanks all! :shipit:

AshishDhama commented on types/store/draft.ts:
```diff
@@ -22,10 +24,10 @@ export type PostDraft = {
 };
+
+export type NewPostDraft = PostDraft & {
+    content?: JSONContent;
```
@michelengelen why do we need to introduce a new type if you add an optional key? I am not sure we are doing version bump for types, but we should update the existing `PostDraft.`
but would leave that decision with web platform cc// @hmhealey

michelengelen replied:
good question ... this is a change that came from the working group during mattercon, so I am not sure who introduced that ... I can just change the existing type, since the comment is valid! :D

AshishDhama commented on tests/setup.js:
```diff
@@ -8,7 +8,7 @@ import '@testing-library/jest-dom';
 import './redux-persist_mock';
 import './react-intl_mock';
-import './react-router-dom_mock';
+import './react-router-dom_mock.ts';
```
I don’t think this is required If I am not wrong we by default search for ts file.

michelengelen replied:
it is needed, since importing `.ts` in `.js` files and vice versa need the extension

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!
